providers/hostingde/types.go

```diff
@@ -0,0 +1,173 @@
+	DNSSECMode   string `json:"dnsSecMode"`
+	EmailAddress string `json:"emailAddress,omitempty"`
+	MasterIP     string `json:"masterIp"`
+	Name         string `json:"name"` // Not required per docs, but required IRL
+	NameUnicode  string `json:"nameUnicode"`
+	// SOAValues    struct {
+	// 	Refresh     uint32 `json:"refresh"`
+	// 	Retry       uint32 `json:"retry"`
+	// 	Expire      uint32 `json:"expire"`
+	// 	TTL         uint32 `json:"ttl"`
+	// 	NegativeTTL uint32 `json:"negativeTtl"`
+	// } `json:"soaValues,omitempty"`
+	Type                  string   `json:"type"`
+	ZoneTransferWhitelist []string `json:"zoneTransferWhitelist"`
+}
+
+type record struct {
+	ID       string `json:"id"`
+	Name     string `json:"name"`
+	Type     string `json:"type"`
+	Content  string `json:"content"`
+	TTL      uint32 `json:"ttl"`
+	Priority uint16 `json:"priority"`
+}
+
+type response struct {
+	Errors   []apiError    `json:"errors"`
+	Response *responseData `json:"response"`
+	Status   string        `json:"status"`
+}
+
+type apiError struct {
+	Code          int    `json:"code"`
+	ContextObject string `json:"contextObject"`
+	ContextPath   string `json:"contextPath"`
+	Text          string `json:"text"`
+	Value         string `json:"value"`
+}
+
+type responseData struct {
+	Data json.RawMessage `json:"data"`
+	Type string          `json:"type"`
+
+	Limit      uint `json:"limit"`
+	Page       uint `json:"page"`
+	TotalPages uint `json:"totalPages"`
+}
+
+func (r *record) nativeToRecord(domain string) *models.RecordConfig {
+	// normalize cname,mx,ns records with dots to be consistent with our config format.
+	if r.Type == "ALIAS" || r.Type == "CNAME" || r.Type == "MX" || r.Type == "NS" || r.Type == "SRV" {
+		if r.Content != "." {
+			r.Content = r.Content + "."
+		}
+	}
+
+	rc := &models.RecordConfig{
+		Type:         "",
+		TTL:          r.TTL,
+		MxPreference: r.Priority,
+		SrvPriority:  r.Priority,
+		Original:     r,
+	}
+	rc.SetLabelFromFQDN(r.Name, domain)
+
+	var err error
+	switch r.Type {
+	case "ALIAS":
+		rc.Type = r.Type
+		rc.Target = r.Content
+	case "NULLMX":
+		err = rc.PopulateFromString("MX", "0 .", domain)
+	case "MX":
+		err = rc.SetTargetMX(uint16(r.Priority), r.Content)
+	case "SRV":
+		err = rc.SetTargetSRVPriorityString(uint16(r.Priority), r.Content)
+	default:
+		if err := rc.PopulateFromString(r.Type, r.Content, domain); err != nil {
+			panic(err)
+		}
+	}
+	if err != nil {
+		panic(err)
+	}
+
+	return rc
+}
+
+func recordToNative(rc *models.RecordConfig) *record {
+	rc.Target = strings.TrimSuffix(rc.Target, ".")
+	record := &record{
+		Name:    rc.NameFQDN,
+		Type:    rc.Type,
+		Content: rc.GetTargetCombined(),
+		TTL:     rc.TTL,
+	}
+
+	switch rc.Type { // #rtype_variations
+	case "A", "AAAA", "ALIAS", "CAA", "CNAME", "DNSKEY", "DS", "NS", "NSEC", "NSEC3", "NSEC3PARAM", "PTR", "RRSIG", "SSHFP", "TSLA", "TXT":
+		// Nothing special.
+	case "MX":
+		record.Priority = rc.MxPreference
+		record.Content = rc.Target
```

Comment:
Rather than rc.Target, please use rc.GetTargetFrield()

For TXT records, please use .TxtStrings (which may require special handling, since it is a []string)

TXT handling will change in a few days.  It might break this code.  See https://github.com/StackExchange/dnscontrol/pull/1063

Reply:
Done. Once you merge the tlim_target_and_txt branch I will create an auditrecords.go.